Add utility to parse a duration string (#560)
stdlib-js/stdlib: lib/node_modules/@stdlib/time/base/parse-duration/README.md

You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# Parse Duration

> Parse a duration string into an object with `minutes`, `seconds`, and `milliseconds`.

<section class="usage">

## Usage

```javascript
var parseDuration = require( '@stdlib/time/base/parse-duration' );
```

#### parseDuration

Parses a duration string such as `1m3s10ms` into an object with `minutes`, `seconds`, and `milliseconds`.

```javascript
var obj = parseDuration( '1m3s10ms' );
// returns { 'minutes': 1, 'seconds': 3, 'milliseconds': 10 }

obj = parseDuration( '1m3s' );
// returns { 'minutes': 1, 'seconds': 3, 'milliseconds': 0 }
```

</section>

<!-- /.usage -->

<!-- Package notes. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="notes">

> -   A duration string is a string containing a sequence of time units. A time unit is a non-negative integer followed by a unit identifier. The following unit identifiers are supported:

Reviewer: ```suggestion
## Notes

-   A duration string is a string containing a sequence of time units. A time unit is a non-negative integer followed by a unit identifier. The following unit identifiers are supported:
```

You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# Parse Duration

> Parse a duration string into an object with `minutes`, `seconds`, and `milliseconds`.

<section class="usage">

## Usage

```javascript
var parseDuration = require( '@stdlib/time/base/parse-duration' );
```

#### parseDuration

Parses a duration string such as `1m3s10ms` into an object with `minutes`, `seconds`, and `milliseconds`.

```javascript
var obj = parseDuration( '1m3s10ms' );
// returns { 'minutes': 1, 'seconds': 3, 'milliseconds': 10 }

obj = parseDuration( '1m3s' );
// returns { 'minutes': 1, 'seconds': 3, 'milliseconds': 0 }
```

</section>

<!-- /.usage -->

<!-- Package notes. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="notes">

## Notes

> -   A duration string is a string containing a sequence of time units. A time unit is a non-negative integer followed by a unit identifier. The following unit identifiers are supported:

Reviewer: IMO we should also support `h` (hours) and `d` (days). Supporting `y` (years) would be harder given the fact that a year is not a fixed amount of time.

Author: Was leaning that way too. Will update!

Author: Changed accordingly.

<!--

@license Apache-2.0

Copyright (c) 2022 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

> # Parse Duration

Reviewer: ```suggestion
# parseDuration
```

> Parse a duration string into an object.

<section class="usage">

## Usage

```javascript
var parseDuration = require( '@stdlib/time/base/parse-duration' );
```

#### parseDuration

> Parses a duration string such as `1m3s10ms` into an object with `days`, `hours`, `minutes`, `seconds`, and `milliseconds`.

Reviewer: ```suggestion
Parses a duration string into an object.
```

```javascript
var obj = parseDuration( '1m3s10ms' );
// returns { 'days': 0, 'hours': 0, 'minutes': 1, 'seconds': 3, 'milliseconds': 10 }

obj = parseDuration( '1m3s' );
// returns { 'days': 0, 'hours': 0, 'minutes': 1, 'seconds': 3, 'milliseconds': 0 }
```

> 

Reviewer: ```suggestion

The returned object has the following properties:

-   **days**: number of days.
-   **hours**: number of hours.
-   **minutes**: number of minutes.
-   **seconds**: number of seconds.
-   **milliseconds**: number of milliseconds.

```

<section class="notes">

## Notes

-   A duration string is a string containing a sequence of time units. A time unit is a non-negative integer followed by a unit identifier. The following unit identifiers are supported:

    -   `d`: days
    -   `h`: hours
    -   `m`: minutes
    -   `s`: seconds
    -   `ms`: milliseconds

    For example, the string `1m3s10ms` is a duration string containing three time units: `1m` (1 minute), `3s` (3 seconds), and `10ms` (10 milliseconds). The string `60m` is a duration string containing a single time unit: `60m` (60 minutes). A duration string must contain at least one time unit. Time units must be supplied in descending order of magnitude (i.e., days, hours, minutes, seconds, milliseconds).

-   Duration strings are case insensitive. For example, the string `1M3S10MS` is equivalent to `1m3s10ms`.

-   If a duration string does not contain a time unit, the respective property is set to `0`.

-   An empty string is considered a valid duration string and is parsed as `0d0h0m0s0ms`.

</section>

<!-- /.notes -->

<section class="examples">

## Examples

> <!-- TODO: better examples -->
> 
> <!-- eslint no-undef: "error" -->

Reviewer: ```suggestion
<!-- eslint no-undef: "error" -->
```

distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# parseDuration

> Parse a duration string into an object.

<section class="usage">

## Usage

```javascript
var parseDuration = require( '@stdlib/time/base/parse-duration' );
```

#### parseDuration

Parses a duration string into an object.

```javascript
var obj = parseDuration( '1m3s10ms' );
// returns { 'days': 0, 'hours': 0, 'minutes': 1, 'seconds': 3, 'milliseconds': 10 }

obj = parseDuration( '1m3s' );
// returns { 'days': 0, 'hours': 0, 'minutes': 1, 'seconds': 3, 'milliseconds': 0 }
```

The returned object has the following properties:

-   **days**: number of days.
-   **hours**: number of hours.
-   **minutes**: number of minutes.
-   **seconds**: number of seconds.
-   **milliseconds**: number of milliseconds.

</section>

<!-- /.usage -->

<!-- Package notes. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="notes">

## Notes

> -   A duration string is a string containing a sequence of time units. A time unit is a non-negative integer followed by a unit identifier. The following unit identifiers are supported:

Reviewer: ```suggestion
-   A duration string is a string containing a sequence of time units. A time unit is a nonnegative integer followed by a unit identifier. The following unit identifiers are supported:
```

    -   `d`: days
    -   `h`: hours
    -   `m`: minutes
    -   `s`: seconds
    -   `ms`: milliseconds

>     For example, the string `1m3s10ms` is a duration string containing three time units: `1m` (1 minute), `3s` (3 seconds), and `10ms` (10 milliseconds). The string `60m` is a duration string containing a single time unit: `60m` (60 minutes). A duration string must contain at least one time unit.

Reviewer: ```suggestion
    For example, the string `1m3s10ms` is a duration string containing three time units: `1m` (1 minute), `3s` (3 seconds), and `10ms` (10 milliseconds). The string `60m` is a duration string containing a single time unit: `60m` (60 minutes).
```

Given that empty strings are considered valid duration strings, this sentence does not seem correct.

Author: Thanks for catching! Suggestion is committed.

>     For example, the string `1m3s10ms` is a duration string containing three time units: `1m` (1 minute), `3s` (3 seconds), and `10ms` (10 milliseconds). The string `60m` is a duration string containing a single time unit: `60m` (60 minutes). A duration string must contain at least one time unit. Time units must be supplied in descending order of magnitude (i.e., days, hours, minutes, seconds, milliseconds).

Reviewer: This currently says that time units must be supplied in descending order of magnitude; however, FWIW, the implementation allows providing them in any order.

Author: Hmm, maybe we should just remove that sentence here then? I don't think it would be worthwhile to enforce the order so maybe we could just make it an implicit assumption.

Reviewer: Can remove from here. Can still include in the `@stdlib/assert/is-duration-string`, if validation enforces a particular order.

Author: Removed.
